### PR TITLE
Connections anti-leak measures

### PR DESCRIPTION
**Pull request by art-gor**
Internal yamux ping functionality enabled to detect dead connections with configured time accuracy,
Message bus corrections,
Message bus for closing connections,
Configure interval for muxed connections to self close after last stream closes

### CODE REVIEW

File: include/libp2p/event/bus.hpp [hunk 1]
```diff
@@ -36,6 +36,7 @@ namespace libp2p::event {
           *first;
         } catch (...) {
           // drop
+          // TODO(artem): log
```
Comment (igor-egorov) [inline include/libp2p/event/bus.hpp:39]: Please log

File: include/libp2p/event/bus.hpp [hunk 2]
```diff
@@ -48,35 +49,47 @@ namespace libp2p::event {
...
      * of this object expires
      */
-    void unsubscribe() {
+    void unsubscribe() noexcept {
       if (handle_.connected()) {
-        handle_.disconnect();
+        try {
+          handle_.disconnect();
+        } catch (...) {
+          // TODO(artem): log
```
Comment (igor-egorov) [inline include/libp2p/event/bus.hpp:65]: Please log

File: include/libp2p/muxer/muxed_connection_config.hpp [hunk 3]
```diff
@@ -21,6 +22,16 @@ namespace libp2p::muxer {
...
     static constexpr size_t kDefaultMaxStreamsNumber = 1000;
     size_t maximum_streams = kDefaultMaxStreamsNumber;
+
+    /// Ping interval for muxers with internal ping feature
+    static constexpr std::chrono::milliseconds kDefaultPingInterval =
+        std::chrono::milliseconds(5000);
+    std::chrono::milliseconds ping_interval = kDefaultPingInterval;
+
+    /// No streams interval after which connection closes itself
+    static constexpr std::chrono::milliseconds kDefaultNoStremsInterval =
```
Comment (igor-egorov) [inline include/libp2p/muxer/muxed_connection_config.hpp:32]:
```suggestion
    static constexpr std::chrono::milliseconds kDefaultNoStreamsInterval =
```

File: include/libp2p/muxer/muxed_connection_config.hpp [hunk 4]
```diff
@@ -21,6 +22,16 @@ namespace libp2p::muxer {
...
+        std::chrono::milliseconds(120000);
+    std::chrono::milliseconds no_streams_interval = kDefaultNoStremsInterval;
```
Comment (igor-egorov) [inline include/libp2p/muxer/muxed_connection_config.hpp:34]:
```suggestion
    std::chrono::milliseconds no_streams_interval = kDefaultNoStreamsInterval;
```

File: include/libp2p/muxer/yamux/yamuxed_connection.hpp [hunk 5]
```diff
@@ -165,15 +168,31 @@ namespace libp2p::connection {
     void onDataWritten(outcome::result<size_t> res, StreamId stream_id,
                        bool some);
 
+    /// Creates new yamux stream
+    std::shared_ptr<Stream> createStream(StreamId stream_id);
+
     /// Erases stream
     void eraseStream(StreamId stream_id);
 
+    /// Erases stream
```
Comment (igor-egorov) [inline include/libp2p/muxer/yamux/yamuxed_connection.hpp:177]: Stream or streams? Not clear from the description how the iterator is going to be used.

File: src/network/impl/connection_manager_impl.cpp [hunk 6]
```diff
@@ -93,25 +60,24 @@ namespace libp2p::network {
...
-      vec.erase(std::remove_if(vec.begin(), vec.end(),
-                               [](auto &&conn) {
-                                 return conn == nullptr || conn->isClosed();
-                               }),
-                vec.end());
+      auto &cs = it->second;
+      for (auto it2 = cs.begin(); it2 != cs.end();) {
+        const auto &conn = *it2;
+        if (conn == nullptr || conn->isClosed()) {
+          // TODO(artem): shold never get here
```
Comment (igor-egorov) [inline src/network/impl/connection_manager_impl.cpp:72]: What is the action required according to this todo? not clear(

File: src/network/impl/connection_manager_impl.cpp [hunk 7]
```diff
@@ -120,14 +86,60 @@ namespace libp2p::network {
...
+    auto it = connections_.find(p);
+    if (it == connections_.end()) {
+      return;
+    }
+
+    auto connections = std::move(it->second);
+    connections_.erase(it);
+
+    if (connections.empty()) {
+      // TODO(artem) log inconsistency
```
Comment (igor-egorov) [inline src/network/impl/connection_manager_impl.cpp:98]: Please put proper logging

**Review by igor-egorov — APPROVED** [review]

File: src/muxer/yamux/yamuxed_connection.cpp [hunk 8]
```diff
@@ -697,9 +713,44 @@ namespace libp2p::connection {
...
+
   void YamuxedConnection::eraseStream(StreamId stream_id) {
     log()->debug("erasing stream {}", stream_id);
     streams_.erase(stream_id);
+    adjustExpireTimer();
+  }
+
+  void YamuxedConnection::erasePendingOutboundStream(
+      PendingOutboundStreams::iterator it) {
+    log()->trace("erasing pending outbound stream {}", it->first);
```
Comment (xDimon) [inline src/muxer/yamux/yamuxed_connection.cpp:733]: Pls, use macros  SL_TRACE, SL_DEBUG, etc.

File: test/libp2p/injector/network_injector_test.cpp [hunk 9]
```diff
@@ -132,7 +178,8 @@ TEST(NetworkBuilder, CustomAdaptorsBuilds) {
   auto network = injector.create<std::shared_ptr<Network> >();
   ASSERT_NE(network, nullptr);
 
-  // upgrader is not created above, because it is not required for any transport
+  // upgrader is not created above, because it is not required for any
+  // transport
```
Comment (xDimon) [inline test/libp2p/injector/network_injector_test.cpp:182]: minor: needless line break in comment

**Review by xDimon — APPROVED** [review]

**Pull request merged**